File: 0116_med_hot_populating-next-right-pointers-in-each-node/solution.java
/*
// Definition for a Node.
*/

import java.util.LinkedList;
import java.util.Queue;

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};


class Solution {

    public Node connect(Node root) {
        if (root == null) {
            return null;
        }

        Queue<Node> queue = new LinkedList<>();

        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            Node preNode = null;

            while (size-- > 0) {
                Node curNode = queue.poll();
                if(curNode.left != null) {
                    queue.offer(curNode.left);
                }
                if(curNode.right != null) {
                    queue.offer(curNode.right);
                }
                if(preNode != null) {
                    preNode.next = curNode;
                }
                preNode = curNode;
            }
        }

        return root;
        
    }
}

File: 0116_med_hot_populating-next-right-pointers-in-each-node/solution.py
"""
# Definition for a Node.
"""

from collections import deque
from typing import Optional


class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
        

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None
        
        queue = deque()
        
        queue.append(root)

        while queue:
            size = len(queue)
            pre_node = None

            for _ in range(size):
                cur_node = queue.popleft()
                if pre_node:
                    pre_node.next = cur_node
                if cur_node.left:
                    queue.append(cur_node.left)
                if cur_node.right:
                    queue.append(cur_node.right)
                pre_node = cur_node
            
        return root
